### DISCUSSION TITLE
v-for not respecting array order

### DISCUSSION THREAD

**Discussion by TomS-** [vuejs/petite-vue#26] [Q&A] [answered]
Having an issue with v-for is not respecting the order of the array but is randomly ordering however it wants. 
```javascript
const venues = PetiteVue.reactive({
    items: [],
    search: "",
    order: "a-z",
    get filtered() {
      let filtered = this.items.filter(item => item.title.toLowerCase().includes(this.search.toLowerCase()));
      filtered = filtered.sort((a, b) => {
        if(this.order == "a-z") {
          if(a.title.toLowerCase() > b.title.toLowerCase()) return -1;
          if(a.title.toLowerCase() < b.title.toLowerCase()) return 1;
          return 0;
        } else if(this.order == "latest") {
          return a.modified - b.modified; 
        }
      });

      console.log( filtered );

      return filtered;
    }
  });

  PetiteVue.createApp({ venues }).mount();
```

Filtered is returning:
![image](https://user-images.githubusercontent.com/4595662/124935649-be752400-dffd-11eb-9a3d-a80e165a0a9f.png)

But the actual order is being displayed as:
![image](https://user-images.githubusercontent.com/4595662/124935826-ebc1d200-dffd-11eb-8e1e-a9d348fe4e99.png)

Does anyone know how to fix this?

**Comment by sqllyw** [comment]
based on your fragment of code, here is a working version, it shows the array in the order you want:
https://jsfiddle.net/sv9oLx0y/

**Reply by TomS-** [reply]
Thanks for this, sorry I should have mentioned. It works correctly until I search. I have updated your jsfiddle with the search function: 
https://jsfiddle.net/tzwb4ecv/

If you type something in search, you will see when it's back to empty it doesn't respect the order of the array you see in console.log.

**Comment by sqllyw** [comment]
Got your point, the official grid.html example has similar issue:

https://jsfiddle.net/gumde7nr/

first pass, you got: Bruce Lee, Chuck Norris, Jackie Chan, Jet Li
type B, then back space, you got Bruce Lee, Jet Li, Jackie Chan, Chuck Norris

A intended behaviour or a bug?

**Reply by TomS-** [reply]
I would assume that this is a bug, but I have no idea what's causing it to not respect the actual array order. It might be some form of caching?

**Comment by nooooooom** [comment]
I think an error occurred when reusing components, I tried to fix it, it looks good at present work.(https://github.com/nooooooom/petite-vue/commit/e2404f65aaf8bd8ce54873321b80d7232ee693c1)

**Reply by go4cas** [reply]
@nooooooom, I am having the same issue. Any idea how/when your changes could be reviewed and merged?

**Reply by nooooooom** [reply]
I don’t know, but just now I discovered that this sorting algorithm has another problem besides the nextBlock lookup error, and I’m fixing it.

**Comment by lumlumi** [comment]
This bug happen only on array update. 

If we take a look in v-for implementation, this is v-for initilisation.

```js
  if (!mounted) {
      blocks = childCtxs.map((s) => mountBlock(s, anchor))
      mounted = true
    } 
```


When updating, you can see below that the index is in reverse order and therefore the html mount is done in reverse. It explains the reverse order of the display. 

```js

      let i = childCtxs.length
      while (i--) {
        const childCtx = childCtxs[i]
        const oldIndex = prevKeyToIndexMap.get(childCtx.key)
        const next = childCtxs[i + 1]
        const nextBlockOldIndex = next && prevKeyToIndexMap.get(next.key)
        const nextBlock =
          nextBlockOldIndex == null ? undefined : blocks[nextBlockOldIndex]
        if (oldIndex == null) {
          // new
          nextBlocks[i] = mountBlock(
            childCtx,
            nextBlock ? nextBlock.el : anchor
          )
        } else {
          // update
          const block = (nextBlocks[i] = blocks[oldIndex])
          Object.assign(block.ctx.scope, childCtx.scope)
          if (oldIndex !== i) {
            // moved
            if (blocks[oldIndex + 1] !== nextBlock) {
              block.insert(parent, nextBlock ? nextBlock.el : anchor)
            }
          }
        }
      }

```

I made a dirty fix to verify this and it's working correctly.  Note that this does not solve the problem with the grid example, which is another issue.

**Reply by lumlumi** [reply]
There is the fix : https://github.com/vuejs/petite-vue/pull/78. Finally it solve grid example issue.

**Comment by Emory-M** [comment]
~~Many pr waiting for, quick fix the bug in min js file, replace `let n=b.length;for(;n--;)` with `for(let n=0;n<b.length;n++)` and self-hosted it.~~

**Reply by nooooooom** [reply]
I think it's not just this kind of problem, because of performance optimization, it will cause the data to be inconsistent with the UI that is finally displayed. You can compare the shuffled array with the UI.
Even's approach is no problem, but there is a situation that when two parallel data move forward, the data comparison will be problematic.

**Reply by nooooooom** [reply]
Reverse in response to `insertBefore`, it is compared to the benefits of positive sequence is more convenient to find the `anchor`, and the ability to quickly know whether they should move even in the case of index changes.I have repaired it on the basis of original thinking, It works now (https://github.com/nooooooom/petite-vue/commit/cc299588330bdec8a17bf1e99938439f1c66d995)

**Reply by lumlumi** [reply]
yes you are right. Can you create PR to trace this ?

**Reply by nooooooom** [reply]
How about this(https://github.com/vuejs/petite-vue/pull/79).

**Comment by TomS-** [comment, marked as answer]
This has been pushed to live now thanks to @nooooooom for the PR.
